msprime/provenance.py

```diff
@@ -188,10 +189,21 @@ def hook(obj):
             elif "__npgeneric__" in obj:
                 return numpy.array([obj["__npgeneric__"]]).astype(obj["dtype"])[0]
             elif "__class__" in obj:
-                module, cls = obj["__class__"].rsplit(".", 1)
+                module, cls = obj.pop("__class__").rsplit(".", 1)
                 module = importlib.import_module(module)
-                del obj["__class__"]
-                return getattr(module, cls)(**obj)
+                cls = getattr(module, cls)
+                if cls == demography.Demography:
+                    # the Demography class normally generates its own population IDs, but
+                    # allow fixed IDs here to ensure they match e.g. IDs in event objects
+                    demography_object = object.__new__(cls)
+                    try:
+                        demography_object.__init__(**obj)
+                    except ValueError as e:
+                        if str(e).startswith("Population ID should not be set"):
```

Comment:
This control flow by exception logic is fragile and distributes the logic of creating Demography objects around different modules. We should add a ``Demography.fromdict`` method that does this, and can be tested independently;.

Reply:
I'm returning to some of my open PRs, and realise that this one has an easy fix, because I just found out that you can pass an InitVar parameter to a dataclass, in this case to say not to worry about making new population IDs in this specific case. I think this is probably the minimally intrusive thing to do.